### ISSUE TITLE
Gaussian MultiReadout assumes shared parameters

### ISSUE THREAD

**Issue by KonstantinWilleke** [sinzlab/neuralpredictors#151]
In the current neuralpredictors main, the `MultipleFullGaussian2d` is inheriting from `MultiReadoutSharedParametersBase`. 
This might be a problem in some cases. For example, if I just want a "normal" MultiGauss2d, without sharing anything, I could just invoke it as it is and it works. 
However, it does not work with `grid_mean_predictor`.
Because, for example, whenever a `grid_mean_predictor` is passed, it expects that there is a source grid. 

The easiest solution, and my proposal, is to make it inherit from `MultiReadoutBase` then, so that the __init__ would look like:

```
class MultipleFullGaussian2d(MultiReadoutBase):
    _base_readout = FullGaussian2d
```

So either MultiReadoutSharedParametersBase is expanded, so it does work with and without sharing, or we have two independent instantiations from the respective base classes. I'm leaning towards the latter.

**Comment by kklurz** [comment]
I'm not sure if I understand the issue correctly. The way I understand it, you are raising two issues that are independent of each other:
- **"The Multireadout-version of the FullGaussian2D readout has to share parameters because it inherits from the `MultiReadoutSharedParametersBase`".** The fact that the multireadout that we call `MultipleFullGaussian2d ` is inheriting from the `MultiReadoutSharedParametersBase` is only for backwards compatibility reasons, so that people can still directly import it and don't get an error. The whole design of the two Multireadout base classes is such that the user would import them both and also import the single base readout and then combine them to their wishes. So the two lines that you gave above are actually the standard use case, conducted in the user's code (we do not want all 2-liner combinations of base readouts and multireadout bases in neuralpredictors itself).
- **Whenever a `grid_mean_predictor` is passed, it expects that there is a source grid.** The grid mean predictor predicts the readout position from a source grid (atm, only a cortical source grid is implemented). If you look at the [code ](https://github.com/sinzlab/neuralpredictors/blob/7f1b3039aeccf04633a52fc32b54889a767c32f6/neuralpredictors/layers/readouts/gaussian.py#L429) of the FullGaussian2D base readout, it is necessary to pass a source grid when `grid_mean_predictor` is not `None`. I'm not sure, how your code could run with passing a `grid_mean_predictor` but no `source_grid`?

**Comment by KonstantinWilleke** [comment]
Based on our discussion in code review, let me rephrase what I mean with this issue.

The state of the current MultiReadouts is as you say:
1) The actual instantiations of the MultiReadouts in `layers.readouts` are there only for backwards compatibility.
2) The user has to define their own MultiReadouts in the respective repo, by choosing either the MultiReadoutBase, or the SharedParametersBase.
3) There are two base classes (with sharing or without), that the user has to choose from.

Why I created this issue is that all of these points can be improved I think to make it more user friendly.

My ideas to improve are the following:
1) We try to combine the two Base classes into one. The default would be to not share parameters, but with the right kwargs, Parameter sharing would be possible. I think that is possible while still being backwards compatible.
2) The `.__init__.py` of the readouts can have all of the MultiReadouts instantiated, with their respective base readouts. So that the user can import everything from there and it's ready to use.
3) Backwards compatibility is still present - the behavior is the same, but the get the "old" MutliReadouts, the users that need the legacy code can still instantiate that in their own repo and import from there.

So taken together: The legacy code moves to the repo of the users that need it (Still with a guarantee that the behavior of this class is fully backwards compatible). And in the new neuralpredictors main, there are all ready to use readouts that can be imported.

**Comment by kklurz** [comment]
The userfriendlynes (is this a word?) in the model function would thus improve from this code 

``` python
from neuralpredictors.layers.readouts import FullGaussian2d, MultiReadoutSharedParametersBase

core = ...
readout = MultiReadoutSharedParametersBase(FullGaussian2d, **readout_parameters)
```

to this code: 

``` python
from neuralpredictors.layers.readouts import MultipleFullGaussian2d

core = ...
readout = MultipleFullGaussian2d(**readout_parameters)
```

The downside of this change would be that we lose the more flexible class-based design by replacing it with a keyword-argument approach:
The reason the class-based design was chosen in the first place is that we wanted the system of different cores, encoders, readouts and multireadouts to be a Lego-like system where one can combine all of these parts according to one's needs (provided that the parts are compatible). At the moment we only have 2 multireadouts which is why the Lego approach might seem a bit overkill but there is no reason to assume that in the future we might not want different kinds of interactions between readouts which would call for a new multireadout. In the current setup, any present or future readout can be combined with any present or future multireadout without the need of retrospective adding of keyword arguments. 

So basically the question boils down to how much we value the flexibility of our package and how important the above described change in usability is to us.

**Comment by KonstantinWilleke** [comment]
In the code review meetings we decided to keep the logic as it is, and go with this one:
```
from neuralpredictors.layers.readouts import FullGaussian2d, MultiReadoutSharedParametersBase

core = ...
readout = MultiReadoutSharedParametersBase(FullGaussian2d, **readout_parameters)
```

The grid mean predictor exclusive to the SharedParametersMultiReadoutBase. Which I agree is reasonable because that is a shared readout parameter between neurons.

https://github.com/sinzlab/neuralpredictors/pull/160
